### ISSUE TITLE
Victor 9000 Disks Don't read On physical machine

### ISSUE THREAD

**Issue by pauldevine** [davidgiven/fluxengine#421]
After successfully writing a Victor 9000 disk to physical media,  I'm afraid I have to report the disk wasn't readable on the actual hardware. When I pop the disk in and just try to do a directory listing I get an error that the disk is unreadable. 

Something is partially recognized, however. When I put the same disk in a cold machine and try to boot from it, the BIOS sees the boot record and starts to boot the machine. It gets partway through the boot process and then errors out with `wrong_sector	equ	05h;	no match on sector - error count expired` So at least the boot record is readable.

**Comment by davidgiven** [comment]
Interesting. The error message doesn't point at a CRC error; instead it's as if it just can't find the sector it's looking for. This could be:

- mismatched format and sector layout table in the boot record
- incorrect encoding of sectors so some aren't being recognised
- er, something else?

Do you have anything like a disk verifier tool, which will read the entire disk on the Victor 9000 and report on what sectors are missing or bad? That would shed more light on what's happening.

**Comment by pauldevine** [comment]
Hi David,
That was a good idea to look at the sectors on the physical machine. The tools I have are all based on looking at the FAT table and then jumping into the sectors from there. I was able to screw around with various disks to get into the sector contents on the disk I wrote with fluxengine. The contents of the sectors looked fine. I then started looking at the header information, and I realized the bit marks for the sector header were shorter on the fluxengine version. I think that's our problem. I'll play with fixing it and see where it takes us.

**Comment by davidgiven** [comment]
Did you get anywhere with this?

According to the documentation (which is known to be wrong), the header ID sequence is really simple, with six sync bytes and then the ID byte. The data ID sequence is five sync bytes and then the ID byte. I see that the _ss format has the correct values but the _ds format doesn't. Which one did you try? If it was the _ds format you might try changing the values of pre_header_sync_bits and pre_data_sync_bits to 60 and 50 respectively in src/formats/victor9k_ds.textpb.

**Comment by pauldevine** [comment]
HI David,
Thanks for the gentle nudge. I did make some small progress. I adjusted the sector headers and was able to get the computer to boot from the written disk, but then it wouldn't subsequently read the darn thing once booted. In noticed the FAT table seemed to be unreadable. I can see there's some small differences in the gaps. The original disk the gap fluxes are actually GCR encoded zeros, while on the one we're writing we're doing just straight binary 0s. I'll continue to play around and should be able to get it together.

**Comment by davidgiven** [comment]
Binary zeroes are definitely wrong as the disk can't actually represent these --- it upsets the automatic head gain in the drive; if the drive doesn't see a pulse, it'll ramp up the gain until it does, and if there aren't any _real_ pulses on the disk you'll just get noise. It's possible that this causes the ID marks to be misread.

But it should be producing 1/0/1/0 patterns:

```
fillBitmapTo(bits, cursor, trackdata.post_index_gap_us() / clockRateUs, { true, false });
```

So I'm not sure what's going on there.

**Comment by pauldevine** [comment]
Maybe I'm misunderstanding the code. In the spot where the Gap I and Gap II values would be, on the Victor 9k-written disk I'm seeing a string of 01010 which I'm interpreting as GCR encoded 0's. In the fluxengine code i'm talking about 
[Victor 9K encoder.cc](https://github.com/davidgiven/fluxengine/tree/master/arch/victor9k/encoder.cc)
line 106:

`write_zero_bits(bits, cursor, trackdata.post_header_gap_bits());`

that looks to my reading that it will not be GCR encoded and end up as a binary 0. Am I understanding the code correctly? When I look at the flux in another tool—I was avoiding using FE to see if it would help me look via another path, I was seeing gaps in the flux pattern that looked like the spots were blank.

I had something that was close to working, kept iterating it into a different state, and now I'm having trouble recreating that intermediate spot that allowed a boot. Stupidly I hadn't been checking in all my changes as I was just playing around. I'm busy the next couple of days but I'll have some time early next week to continue iterating. It actually feels relatively close.

**Comment by davidgiven** [comment]
Oops! You're absolutely right. I was looking at the wrong gap --- see line 175. Yes, that's completely wrong. I have no idea why I did that.

Try changing the `write_zero_bits()` calls to `fillBitmapTo()`:

```
fillBitmapTo(bits, cursor, trackdata.post_header_gap_bits(), {true, false});
```

The alternative 1010... pattern is actually interpreted as alternate 01010 10101 groupings, which is GCR encoded 0 and f, but that should be fine?

I'm sorry you're having to do the work here, but without a machine I can't test  the results...

**Comment by pauldevine** [comment]
I replaced the write_zero_bits() with the fillBitmapTo() and it seemed to help. I also adjusted some of the lengths to more closely match what I think are the correct values. 

I'm now back in the same spot I described earlier, a disk inserted seems to not work, with unable to find sector errors. But when I try a cold boot the machine does partially boot and then craps out at the stage it would normally load the device drivers. I made a video of it so you could see what I'm talking about. 

[Normal Boot](https://drive.google.com/file/d/1gxaZZURD-gQ0D_6RQGC5RnEbyjW6mEvF/view?usp=sharing)
[Boot of FluxEngine written disc](https://drive.google.com/file/d/1iHnIXfNNUIRBSfTRLJj3J92t8WSh5cdx/view?usp=sharing)

You can see the normal boot seems to read more quickly, I think it's reading the track more than once on the FluxEngine-sourced disk. I'll continue to play with the sync header lengths and the gap values and see if I can figure out what's happening. The code that achieved the above is this branch https://github.com/pauldevine/fluxengine/commit/51df0aa799a3ba473277227bbc1e06760e22c0cb

**Comment by pauldevine** [comment]
Hi David,
I can report good progress. @Fozztexx [posted today](https://twitter.com/FozzTexx/status/1496164885608079366) on twitter about writing a Victor9k disk. The conversation there made me realize I've been trying to debug the fluxengine writing just by altering the bitstream on the disk, and I haven't been adjusting the timing at all. After playing around with both the timing and the formatting, I was able to get this branch https://github.com/pauldevine/fluxengine/tree/victor9k_partial to successfully write a disk that will boot the machine and the rest of the files and FAT table appear to work fine. I used diskcopy to make a successful copy of the disk, which also booted and seemed to work. Diskcopy was the only tool I knew that would sequentially read each sector of the disk.
 
I'm not yet submitting a PR as in my many iterations through playing with the encoder.cc I broke the tests that validate reading/writing. The decoder and encoder are no longer in sync. I'll fix that later in the week. I won't have time to play for a few days. I also have so far only tested a single-sided disk. I'll validate on a double-sided as well.

Thanks for all your help. I feel like I just rescued some old machines out there.

Paul
